File: src/security/passwordAnalyzer.js
var PasswordAnalyzer = function (password) {
  this._password = password;
  this._score = 0;
  this._rules = [];
  this._history = {};
  this._typesFound = new Set();
  this._analyze();
};

PasswordAnalyzer.prototype._analyze = function () {
  this._rules = [];
  this._score = 0;
  if (!this._password || this._password.length === 0) {
    this._rules.push("Password is empty");
    return;
  }
  this._typesFound = new Set();
  for (var i = 0; i < this._password.length; i++) {
    var c = this._password[i];
    if (c >= 'A' && c <= 'Z') this._typesFound.add("UPPER");
    else if (c >= 'a' && c <= 'z') this._typesFound.add("LOWER");
    else if (c >= '0' && c <= '9') this._typesFound.add("DIGIT");
    else this._typesFound.add("SPECIAL");
  }
  if (this._password.length >= 8) { this._rules.push("Min 8 characters"); this._score++; } else { this._rules.push("Min 8 characters"); }
  if (this._password.length >= 12) { this._rules.push("Min 12 characters"); this._score++; } else { this._rules.push("Min 12 characters"); }
  if (this._password.length >= 16) { this._rules.push("Min 16 characters"); this._score++; } else { this._rules.push("Min 16 characters"); }
  if (/[a-z]/.test(this._password)) { this._rules.push("Has lowercase letter"); this._score++; } else { this._rules.push("Has lowercase letter"); }
  if (/[A-Z]/.test(this._password)) { this._rules.push("Has uppercase letter"); this._score++; } else { this._rules.push("Has uppercase letter"); }
  if (/[0-9]/.test(this._password)) { this._rules.push("Has number"); this._score++; } else { this._rules.push("Has number"); }
  if (/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(this._password)) { this._rules.push("Has special character"); this._score++; } else { this._rules.push("Has special character"); }
  if (this._typesFound.size >= 3) { this._rules.push("3+ character types"); this._score++; } else { this._rules.push("3+ character types"); }
  if (this._typesFound.size >= 4) { this._rules.push("All 4 character types"); this._score++; } else { this._rules.push("All 4 character types"); }
  var noRepeat = true;
  for (var j = 1; j < this._password.length; j++) {
    if (this._password[j] === this._password[j - 1]) { noRepeat = false; break; }
  }
  if (noRepeat) { this._rules.push("No repeated consecutive chars"); this._score++; } else { this._rules.push("No repeated consecutive chars"); }
  this._history[this._password] = this._score;
};

PasswordAnalyzer.prototype.getScore = function () { return this._score; };
PasswordAnalyzer.prototype.getRules = function () { return this._rules; };
PasswordAnalyzer.prototype.getHistory = function () { return this._history; };
PasswordAnalyzer.prototype.getTypesFound = function () { return this._typesFound; };
PasswordAnalyzer.prototype.getStrengthLabel = function () {
  if (this._score <= 3) return "Weak";
  if (this._score <= 6) return "Medium";
  return "Strong";
};

File: java/src/PasswordAnalyzer.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

public class PasswordAnalyzer {
    private String _password;
    private int _score;
    private ArrayList<String> _rules;
    private HashMap<String, Integer> _history;
    private HashSet<String> _typesFound;

    public PasswordAnalyzer(String password) {
        this._password = password;
        this._rules = new ArrayList<>();
        this._history = new HashMap<>();
        this._typesFound = new HashSet<>();
        analyze();
    }

    private void analyze() {
        _rules.clear();
        _score = 0;

        if (_password == null || _password.isEmpty()) {
            _rules.add("Password is empty");
            return;
        }

        _typesFound.clear();
        for (char c : _password.toCharArray()) {
            if (Character.isUpperCase(c)) _typesFound.add("UPPER");
            else if (Character.isLowerCase(c)) _typesFound.add("LOWER");
            else if (Character.isDigit(c)) _typesFound.add("DIGIT");
            else _typesFound.add("SPECIAL");
        }

        if (_password.length() >= 8) { _rules.add("Min 8 characters"); _score++; }
        else { _rules.add("Min 8 characters"); }

        if (_password.length() >= 12) { _rules.add("Min 12 characters"); _score++; }
        else { _rules.add("Min 12 characters"); }

        if (_password.length() >= 16) { _rules.add("Min 16 characters"); _score++; }
        else { _rules.add("Min 16 characters"); }

        if (_password.matches(".*[a-z].*")) { _rules.add("Has lowercase letter"); _score++; }
        else { _rules.add("Has lowercase letter"); }

        if (_password.matches(".*[A-Z].*")) { _rules.add("Has uppercase letter"); _score++; }
        else { _rules.add("Has uppercase letter"); }

        if (_password.matches(".*[0-9].*")) { _rules.add("Has number"); _score++; }
        else { _rules.add("Has number"); }

        if (_password.matches(".*[!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>\\/?].*")) {
            _rules.add("Has special character"); _score++;
        } else {
            _rules.add("Has special character");
        }

        if (_typesFound.size() >= 3) { _rules.add("3+ character types"); _score++; }
        else { _rules.add("3+ character types"); }

        if (_typesFound.size() >= 4) { _rules.add("All 4 character types"); _score++; }
        else { _rules.add("All 4 character types"); }

        boolean noRepeat = true;
        for (int i = 1; i < _password.length(); i++) {
            if (_password.charAt(i) == _password.charAt(i - 1)) {
                noRepeat = false;
                break;
            }
        }
        if (noRepeat) { _rules.add("No repeated consecutive chars"); _score++; }
        else { _rules.add("No repeated consecutive chars"); }

        _history.put(_password, _score);
    }

    public int getScore() { return _score; }

    public ArrayList<String> getRules() { return _rules; }

    public HashMap<String, Integer> getHistory() { return _history; }

    public HashSet<String> getTypesFound() { return _typesFound; }

    public String getStrengthLabel() {
        if (_score <= 3) return "Weak";
        if (_score <= 6) return "Medium";
        return "Strong";
    }
}
